### ISSUE TITLE
SpecifyShape is a duplicate of assert_shape?

### ISSUE THREAD

**Issue by brandonwillard** [pymc-devs/Theano-PyMC#269]
It really looks like `theano.compile.ops.SpecifyShape` provides the same functionality as `theano.tensor.nnet.assert_shape`.

If so, we should remove one of them.

**Comment by hs2361** [comment]
@brandonwillard Are you referring to `aesara.tensor.shape.SpecifyShape` and `aesara.tensor.nnet.abstract_conv.assert_shape` ?

**Comment by brandonwillard** [comment]
> Are you referring to `aesara.tensor.shape.SpecifyShape` and `aesara.tensor.nnet.abstract_conv.assert_shape` ?

That sounds correct, since some modules and sub-packages were refactored since this issue was created.

**Comment by hs2361** [comment]
I went through the code in `aesara.tensor.shape.SpecifyShape.infer_shape`, however I don't see how it can be a replacement for what `aesara.tensor.nnet.abstract_conv.assert_shape` does. The latter wraps `x` into an `Assert`, however, the former simply raises an `AssertionError`, only if the shapes don't match. Could you clarify a little @brandonwillard ?

**Comment by brandonwillard** [comment]
Yes, the `Assert`s created by `assert_shape` will simply raise `AssertionError`s if the actual shape and the expected shape don't match (at `Op.perform`/run/evaluation-time), and although `SpecifyShape` will also do the same&mdash;albeit not by way of `Assert`&mdash;it has the added advantage of providing shape information via its `SpecifyShape.infer_shape`.  

In other words, `SpecifyShape` does what `assert_shape` does, but also assists the `ShapeFeature` in determining a `Variable`'s shape.  If we're going to remove one of them, it should be the trivial and much less functional `assert_shape`.

**Comment by hs2361** [comment]
Alright then, I'll work on removing `assert_shape`. I don't think it is being used anywhere other than in a test, `tests.tensor.nnet.test_abstract_conv`, so it wouldn't need to be removed from anywhere else, other than from that test.

**Comment by hs2361** [comment]
@brandonwillard I've opened a draft PR (#308). In the PR, I've replaced the functionality of the assert_shape function with an AssertionError containing the appropriate message. However, I'm not exactly sure if this is what you meant. Secondly, the tests in `tests/tensor/nnet/test_abstract_conv.py` will fail, because I am raising AssertionErrors, and the tests don't expect these errors (earlier, the object would have been wrapped in an `Assert`, and this would have raised an `AssertionError` later on). I wasn't sure if the tests were to be changed as well, since 19 of them would need to be changed. 

Please have a look at my changes, and do let me know as to how I can proceed further.

**Comment by ricardoV94** [comment]
`assert_shape` allowed for partial shape via `None`, which `SpecifyShape` could not do before. 

This should be fixed after #825, and then we can remove this helper function afterwards
